packages/front-end/components/Settings/ClickToReveal.tsx

```diff
@@ -1,59 +1,111 @@
-import { ReactElement, ReactNode, useState } from "react";
-import { FaExclamationTriangle } from "react-icons/fa";
-import LoadingSpinner from "../LoadingSpinner";
+import { SecretApiKey } from "back-end/types/apikey";
+import clsx from "clsx";
+import { useState } from "react";
+import { FaExclamationTriangle, FaEye, FaEyeSlash } from "react-icons/fa";
+import { useAuth } from "../../services/auth";
 import Tooltip from "../Tooltip";
 
+export type SecretKey = {
+  [key: string]: string;
+};
 export interface Props {
-  valueWhenHidden: ReactNode;
-  getValue: () => Promise<string>;
-  children: (value: string) => ReactElement;
+  rowReverse?: boolean;
+  keyId: string;
+  currentCopiedString: string;
+  setCurrentCopiedString: (value: string) => void;
 }
 
-export default function ClickToReveal({
-  valueWhenHidden,
-  getValue,
-  children,
+export default function RevealHiddenKey({
+  keyId,
+  rowReverse,
+  currentCopiedString,
+  setCurrentCopiedString,
 }: Props) {
-  const [value, setValue] = useState("");
-  const [loading, setLoading] = useState(false);
+  const { apiCall } = useAuth();
   const [error, setError] = useState("");
+  const [revealedSecretKey, setRevealedSecretKey] = useState<SecretKey | null>(
+    {}
+  );
 
-  if (value) {
-    return children(value);
-  }
-
-  if (loading) {
-    return (
-      <span>
-        <LoadingSpinner /> loading...
-      </span>
-    );
-  }
+  const hidden = !revealedSecretKey || !revealedSecretKey[keyId];
 
   return (
-    <a
-      href="#"
-      onClick={async (e) => {
-        e.preventDefault();
-        setLoading(true);
-        setError("");
-
-        try {
-          const actualValue = await getValue();
-          setValue(actualValue);
-        } catch (e) {
-          console.error(e);
-          setError(e.message);
-        }
-        setLoading(false);
-      }}
+    <div
+      className={clsx("d-flex", rowReverse ? "flex-row-reverse" : "flex-row")}
     >
-      {valueWhenHidden}{" "}
+      <span
+        role="button"
+        onClick={async () => {
+          if (hidden) {
+            try {
+              const res = await apiCall<{ key: SecretApiKey }>(`/keys/reveal`, {
+                method: "POST",
+                body: JSON.stringify({
+                  id: keyId,
+                }),
+              });
+              setRevealedSecretKey({
+                [keyId]: res.key.key,
+              });
+            } catch (e) {
+              setError(e.message);
+            }
+          } else {
+            setRevealedSecretKey(null);
+          }
+        }}
```

Comment:
I would memoize this with `useCallback` so it's not recreated with every component render.

Reply:
@tinahollygb I did a pretty substantial refactor of where we're fetching this data, it's now defined within the parent component and passed as a prop to `ClickToReveal` - in doing so, I also updated the logic so we're not re-fetching the same data multiple times. With the new structure, I don't think it makes sense to consume the `useCallback` hook anymore. However, I would love for you to double check me on that. Thanks!